auth/src/main/java/com/firebase/ui/auth/util/data/ProviderUtils.java

```diff
@@ -154,41 +155,60 @@ public static String accountTypeToProviderId(@NonNull String accountType) {
         return config;
     }
 
-    public static Task<String> fetchTopProvider(FirebaseAuth auth, @NonNull String email) {
+    public static Task<List<String>> fetchSortedProviders(@NonNull FirebaseAuth auth,
+                                                          @NonNull final FlowParameters params,
+                                                          @NonNull String email) {
         if (TextUtils.isEmpty(email)) {
             return Tasks.forException(new NullPointerException("Email cannot be empty"));
         }
 
         return auth.fetchSignInMethodsForEmail(email)
-                .continueWith(new Continuation<SignInMethodQueryResult, String>() {
+                .continueWith(new Continuation<SignInMethodQueryResult, List<String>>() {
                     @Override
-                    public String then(@NonNull Task<SignInMethodQueryResult> task) {
-                        if (!task.isSuccessful()) return null;
-
+                    public List<String> then(@NonNull Task<SignInMethodQueryResult> task) {
                         List<String> methods = task.getResult().getSignInMethods();
-                        return methods == null || methods.isEmpty()
-                                ? null : methods.get(methods.size() - 1);
+                        if (methods == null) { methods = new ArrayList<>(); }
+
+                        List<String> allowedProviders = new ArrayList<>(params.providers.size());
+                        for (AuthUI.IdpConfig provider : params.providers) {
+                            allowedProviders.add(provider.getProviderId());
+                        }
+
+                        List<String> lastSignedInProviders = new ArrayList<>(methods.size());
+                        for (String method : methods) {
+                            String id = signInMethodToProviderId(method);
+                            if (allowedProviders.contains(id)) {
+                                lastSignedInProviders.add(0, id);
+                            }
+                        }
+
+                        if (lastSignedInProviders.contains(GoogleAuthProvider.PROVIDER_ID)) {
```

Comment:
Add a comment here explaining what you're doing (making sure Google is always first) and why?